### PR TITLE
Feature/fast materialize/simba

### PR DESCRIPTION
**Pull request by simba-git**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: provider/dynamodb.go [hunk 1]
```diff
@@ -801,7 +850,10 @@ func castNumberToFloat32(value any) (float32, error) {
 		return float32(typed), nil
 	case string:
 		f64, err := strconv.ParseFloat(typed, 32)
-		return float32(f64), err
+		if err != nil {
+			return 0, fmt.Errorf("Type error: Expected numerical type and got %T", typed)
+		}
```
Comment (epps) [inline provider/dynamodb.go:855]: Do you mean to use `fferr` here?

File: provider/dynamodb.go [hunk 2]
```diff
@@ -431,6 +443,43 @@ func (store *dynamodbOnlineStore) GetImport(id ImportID) (Import, error) {
...
+	for i, serItem := range serialized {
+		reqs[i] = types.WriteRequest{PutRequest: &types.PutRequest{Item: serItem}}
+	}
+	batchInput := &dynamodb.BatchWriteItemInput{
+		RequestItems: map[string][]types.WriteRequest{
+			table.key.ToTableName(): reqs,
+		},
+	}
+	if _, err := table.client.BatchWriteItem(context.TODO(), batchInput); err != nil {
+		return err
```
Comment (epps) [inline provider/dynamodb.go:474]: Do you mean to use `fferr` here given we'd be returning the client's error instead of ours?

File: provider/dynamodb.go [hunk 3]
```diff
@@ -825,7 +877,11 @@ func castNumberToFloat64(value any) (float64, error) {
 	case float64:
 		return typed, nil
 	case string:
-		return strconv.ParseFloat(typed, 64)
+		f64, err := strconv.ParseFloat(typed, 64)
+		if err != nil {
+			return 0, fmt.Errorf("Type error: Expected numerical type and got %T", typed)
+		}
```
Comment (epps) [inline provider/dynamodb.go:883]: Do you mean to use `fferr` here, too?

File: provider/filestore.go [hunk 4]
```diff
@@ -976,12 +1004,12 @@ func (store *genericFileStore) Write(path filestore.Filepath, data []byte) error
...
 		func() error {
 			blob, errRetr := store.bucket.ReadAll(ctx, path.Key())
-			fmt.Printf("Read (%d) bytes from bucket (%s) after write\n", len(data), path.Key())
 			if errRetr != nil {
-				return re.Unrecoverable(errRetr)
+				return errRetr
 			} else if !bytes.Equal(blob, data) {
 				return fmt.Errorf("blob read from bucket does not match blob written to bucket")
 			}
+			fmt.Printf("Read (%d) bytes from bucket (%s) after write\n", len(blob), path.Key())
```
Comment (epps) [inline provider/filestore.go:1012]: Do you want to convert this to a `Debugw` log?

Comment (simba-git) [reply provider/filestore.go:1012]: Will leave for Riddhi to figure out

File: provider/dynamodb.go (hunk 1 above)
Comment (simba-git) [reply provider/dynamodb.go:855]: these get wrapped in the method above

File: provider/dynamodb.go [hunk 5]
```diff
@@ -327,6 +338,7 @@ func (store *dynamodbOnlineStore) CheckHealth() (bool, error) {
 	return true, nil
 }
 
+// TODO(simba) Make this work with Serialize V1
```
Comment (anthonylasso) [inline provider/dynamodb.go:341]: todo?

File: provider/k8s.go [hunk 6]
```diff
@@ -1175,6 +1175,7 @@ func (mat FileStoreMaterialization) NumRows() (int64, error) {
 	if err != nil {
 		return 0, err
 	}
+	// TODO(simba) THIS IS INCORRECT!
```
Comment (anthonylasso) [inline provider/k8s.go:1178]: todo?

File: provider/k8s.go [hunk 7]
```diff
@@ -1206,6 +1207,7 @@ func (mat FileStoreMaterialization) IterateSegment(begin, end int64) (FeatureIte
 	}
 	i := int64(0)
 	for i = 0; i < begin; i++ {
+		// TODO(simba) Can't actually skip the error here
```
Comment (anthonylasso) [inline provider/k8s.go:1210]: error + todo

File: provider/k8s.go [hunk 8]
```diff
@@ -1215,17 +1217,69 @@ func (mat FileStoreMaterialization) IterateSegment(begin, end int64) (FeatureIte
 	}, nil
 }
 
+func (mat FileStoreMaterialization) NumChunks() (int, error) {
+	searchPath, err := mat.store.CreateFilePath(fileStoreResourcePath(mat.id))
+	if err != nil {
+		return -1, err
```
Comment (anthonylasso) [inline provider/k8s.go:1223]: convert -1 to a const. easy to fat finger a diff value between the 8 uses.

File: provider/offline.go [hunk 9]
```diff
@@ -28,6 +28,9 @@ import (
 	"golang.org/x/text/language"
 )
 
+// defaultRowsPerChunk is the number of rows in a chunk when using materializations.
+const defaultRowsPerChunk int64 = 100000
```
Comment (anthonylasso) [inline provider/offline.go:32]: nit: use `100_000` to make it more legible.

File: provider/offline.go [hunk 10]
```diff
@@ -953,24 +964,44 @@ func (table *memoryOfflineTable) WriteBatch(recs []ResourceRecord) error {
 	return nil
 }
 
-type memoryMaterialization struct {
-	id   MaterializationID
-	data []ResourceRecord
+// Used in runner/copy_test.go
+type MemoryMaterialization struct {
+	// Lowercase do so it doesn't name thrash ID()
```
Comment (anthonylasso) [inline provider/offline.go:969]: ??

File: coordinator/coordinator_test.go [hunk 11]
```diff
@@ -1350,17 +1337,25 @@ func testCoordinatorTrainingSet(addr string) error {
...
-		fmt.Printf("waiting for job %v to be deleted\n", tsID)
 	}
-	if elapsed >= time.Duration(10)*time.Second {
+	if elapsed >= waitTime {
 		return fmt.Errorf("timed out waiting for job to delete")
 	}
 	ts_complete, err := client.GetTrainingSetVariant(ctx, metadata.NameVariant{Name: tsName, Variant: ""})
 	if err != nil {
-		return fmt.Errorf("could not get training set variant")
+		return fmt.Errorf("could not get training set variant: %v", err)
```
Comment (aolfat) [inline coordinator/coordinator_test.go:1358]: should we fferror it

File: provider/filestore.go [hunk 12]
```diff
@@ -270,6 +277,27 @@ func NewS3FileStore(config Config) (FileStore, error) {
 		return nil, wrapped
 	}
 
+	args := []func(*awsv2cfg.LoadOptions) error{
```
Comment (aolfat) [inline provider/filestore.go:280]: what is all this new stuff for? so we can connect with retries?

File: runner/copy.go [hunk 13]
```diff
@@ -258,39 +282,64 @@ func (m *MaterializedChunkRunnerConfig) Deserialize(config Config) error {
...
-		return nil, err
-	}
-	offlineProvider, err := provider.Get(runnerConfig.OfflineType, runnerConfig.OfflineConfig)
-	if err != nil {
-		return nil, err
-	}
-	onlineStore, err := onlineProvider.AsOnlineStore()
-	if err != nil {
-		return nil, err
+	cachedOnline, found := onlineProviderCache.Load(onlineCfgStr)
```
Comment (aolfat) [inline runner/copy.go:303]: I wonder if it makse sense to move the caching logic into provider.Get

Comment (aolfat) [reply runner/copy.go:303]: @sdreyer did some optimization work for redis I wonder if any of that work applies

File: provider/filestore.go (hunk 12 above)
Comment (aolfat) [reply provider/filestore.go:280]: just a thought, this looks like code I would copy from here and paste in a scratch file to connect to s3 to test out their api's locally, might be good to have it as a separate function so folks can easily search/use it.

File: provider/dynamodb.go (hunk 5 above)
Comment (simba-git) [reply provider/dynamodb.go:341]: Left on purpose

File: provider/k8s.go (hunk 6 above)
Comment (simba-git) [reply provider/k8s.go:1178]: Old code

File: provider/k8s.go (hunk 7 above)
Comment (simba-git) [reply provider/k8s.go:1210]: Old code

File: provider/k8s.go (hunk 8 above)
Comment (simba-git) [reply provider/k8s.go:1223]: Yeah but its actually the error that should be checked, the -1 is to hopefully force some sort of error if you forget to check the err

Comment (simba-git) [reply provider/k8s.go:1223]: If this was a special value that meant error, I'd agree, but I just arbritrarily decided -1 instead of 0

File: provider/offline.go (hunk 10 above)
Comment (simba-git) [reply provider/offline.go:969]: I'll make clearer, it can't be named Id because there is already an ID method on this struct

File: runner/copy.go (hunk 13 above)
Comment (simba-git) [reply runner/copy.go:303]: I actually turned this off, because it breaks all tests and has really gnarly edge cases where you can get two different providers with the same config

File: provider/filestore.go (hunk 12 above)
Comment (simba-git) [reply provider/filestore.go:280]: I want to move all the connection stuff into one place soon, makes testing easier too

**Review by sdreyer — APPROVED** [review]

File: provider/dynamodb.go [hunk 14]
```diff
@@ -431,6 +443,43 @@ func (store *dynamodbOnlineStore) GetImport(id ImportID) (Import, error) {
 	return S3Import{id: id, status: string(output.ImportTableDescription.ImportStatus), errorMessage: errorMessage}, nil
 }
 
+const maxDynamoBatchSize = 25
```
[the remark this answers is not in the document]
Comment (simba-git) [reply provider/dynamodb.go:446]: dynamo limit

File: provider/k8s.go (hunk 8 above)
Comment (anthonylasso) [reply provider/k8s.go:1223]: it's less about the error checking logic and more on the syntax consistency of using a sentinel value in multiple spots. Like if someone fat fingers `-11` then that could cause a bug if a conditional explicitly checks for -1

ex. `if iter.maxIdx != -1` -> some random explosion or data bug

but if you think it can't happen then you're good to resolve!

Comment (aolfat) [reply provider/k8s.go:1223]: I think what Simba says makes sense, the  value doesn't matter if there's an error

Comment (simba-git) [reply provider/k8s.go:1223]: Yeah but shouldn't depend on that value at all, should check the error. I see your point, but I think having a const and exporting it is overkill since it'd have to be documented and would make the behavior fixed.
